keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1187 @@
+-->
+
+The proposed design involves extending CSI with the `VolumeSnapshotDelta` CRD.
+Storage providers can opt in to support this feature by implementing the
+`VOLUME_SNAPSHOT_DELTA_SERVICE` capability in their CSI drivers.
+
+The `VolumeSnapshotDelta` resource is a namespace-scoped resource. It must be
+created in the same namespace as the base and target CSI `VolumeSnapshot`s.
+
+### CBT Datapath Worklow
+
+A Kubernetes user (can be a human or a computer process) initiates the datapath
+workflow by creating a new `VolumeSnapshotDelta` custom resource. The
+`VolumeSnapshotDelta` controller, deployed as a CSI sidecar in the storage
+provider's CSI driver, watches for new `VolumeSnapshotDelta` resources:
+
+![CBT Step 1](./img/cbt-step-01.png)
+
+Instead of satisfying the request directly, the controller constructs a callback
+URL and adds it to the resource's `status` subresource. The user can then use
+the callback URL to fetch the list of changed block metadata:
+
+![CBT Step 2](./img/cbt-step-02.png)
+
+The callback URL points to the `VolumeSnapshotDelta` listener, which runs in
+the same container as the `VolumeSnapshotDelta` controller. The CSI driver
+`Service` resource's FQDN will be used as the hostname of the callback URL, which
+essentially looks like:
+
+```sh
+https://<csi-driver.ns.svc.cluster.local>:<listener-port>/<resource-namespace>/<resource-name>
+```
+
+The `VolumeSnapshotDelta` listener delegates the authentication and
+authorisation of the request to the Kubernetes API server via the `TokenReview`
+and `SubjectAccessReview` APIs, in order to protect both the `nonResourceURL`
+endpoint as well as access to the `VolumeSnapshotDelta` resources.
+
+The user must include the `Authorization` header in the request, using an
+authorised service account's secret token as the bearer token. If the user's
+service account is deployed with `automountServiceAccountToken` set to `false`,
+they will have to extract the appropriate token from their secret.
+
+The `VolumeSnapshotDelta` listener then issues a GRPC call to the
+`GetVolumeSnapshotDelta` service on the storage provider's CSI plugin sidecar:
+
+![CBT Step 3](./img/cbt-step-03.png)
+
+The original `VolumeSnapshotDelta` resource can be retrieved by using the
+identifier (namespace and name) embedded in the URL path, in case there are
+user-provided input parameters which need to be included in the GRPC request.
+
+It is the storage provider's CSI plugin responsibility to call the provider's
+CBT endpoint and manage the in-between authentication and authorisation
+protocols.
+
+The response payloads are then directly returned to the user from the
+`VolumeSnapshotDelta` listener. This synchronous request/response
+mechanism removes the needs to persist the response payloads in etcd.
+
+Pagination parameters from the storage provider will also be included in the
+listener's response to the user. The user will be responsible for coordinating
+subsequent paginated requests.
```

Comment:
There should be an example - how the response body will look like? JSON? With what fields?